source/Octopus.Tentacle/Kubernetes/Diagnostics/PersistenceProvider.cs

```diff
@@ -0,0 +1,50 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using k8s.Models;
+using Octopus.Diagnostics;
+
+namespace Octopus.Tentacle.Kubernetes.Diagnostics
+{
+    public interface IPersistenceProvider
+    {
+        string? GetValue(string key);
+        void PersistValue(string key, string value);
+        ImmutableDictionary<string, string> ReadValues();
+    }
+
+    public class PersistenceProvider : IPersistenceProvider
+    {
+        public delegate PersistenceProvider Factory(string configMapName);
+        
+        readonly string configMapName;
+        readonly IKubernetesConfigMapService configMapService;
+        readonly Lazy<V1ConfigMap> metricsConfigMap;
+        IDictionary<string, string> ConfigMapData => metricsConfigMap.Value.Data ??= new Dictionary<string, string>();
+
+        public PersistenceProvider(string configMapName, IKubernetesConfigMapService configMapService)
+        {
+            this.configMapService = configMapService;
+            this.configMapName = configMapName;
+            metricsConfigMap = new Lazy<V1ConfigMap>(() => configMapService.TryGet(this.configMapName, CancellationToken.None).GetAwaiter().GetResult()
+                ?? throw new InvalidOperationException($"Unable to retrieve Tentacle Configuration from config map for namespace {KubernetesConfig.Namespace}"));
```

Comment:
Maybe in a new PR but I think this should be an `AsyncLazy<>` of some kind. There is one implemented in Server - we can probably copy+paste and use it.

This would cause all the methods to become async though... But I think it's better to avoid the `GetAwaiter().GetResult()` if we can.

Reply:
I'll raise a ticket to do - but atm this is the same as the other configmap usage.

Reply:
Ah fair enough then 👍